Purchases/Public/RCPurchases.m

```diff
@@ -163,88 +210,235 @@ - (void)addAttributionData:(NSDictionary *)data
+        [self updateCachesWithCompletionBlock:^(RCPurchaserInfo *info, NSError *error) {
+            if (info) {
+                [self sendUpdatedPurchaserInfoToDelegateIfChanged:info];
+            }
+        }];
+    }
 }
 
-- (void)readPurchaserInfoFromCache {
-    NSData *purchaserInfoData = [self.userDefaults dataForKey:self.purchaserInfoUserDefaultCacheKey];
+- (RCPurchaserInfo *)readPurchaserInfoFromCache {
+    NSData *purchaserInfoData = [self.userDefaults dataForKey:[self purchaserInfoUserDefaultCacheKeyForAppUserID:self.appUserID]];
     if (purchaserInfoData) {
         NSError *jsonError;
         NSDictionary *infoDict = [NSJSONSerialization JSONObjectWithData:purchaserInfoData options:0 error:&jsonError];
         if (jsonError == nil && infoDict != nil) {
             RCPurchaserInfo *info = [[RCPurchaserInfo alloc] initWithData:infoDict];
-            if (info) {
-                [self handleUpdatedPurchaserInfo:info error:nil];
-            }
+            return info;
         }
     }
-}
-
-- (void)cachePurchaserInfo:(RCPurchaserInfo *)info {
-    if (info.JSONObject) {
-        NSError *jsonError = nil;
-        NSData *jsonData = [NSJSONSerialization dataWithJSONObject:info.JSONObject
-                                                           options:0
-                                                             error:&jsonError];
-        if (jsonError == nil) {
-            [self.userDefaults setObject:jsonData
-                                  forKey:self.purchaserInfoUserDefaultCacheKey];
-        }
+    return nil;
+}
+
+- (void)cachePurchaserInfo:(RCPurchaserInfo *)info forAppUserID:(NSString *)appUserID {
+    if (info) {
+        [self dispatch:^{
+            if (info.JSONObject) {
+                NSError *jsonError = nil;
+                NSData *jsonData = [NSJSONSerialization dataWithJSONObject:info.JSONObject
+                                                                   options:0
+                                                                     error:&jsonError];
+                if (jsonError == nil) {
+                    [self.userDefaults setObject:jsonData
+                                          forKey:[self purchaserInfoUserDefaultCacheKeyForAppUserID:appUserID]];
+                }
+            }
+        }];
     }
 }
 
-- (void)updateCaches
-{
+- (BOOL)isCacheStale {
     NSTimeInterval timeSinceLastCheck = -[self.cachesLastUpdated timeIntervalSinceNow];
-    if (self.cachesLastUpdated != nil && timeSinceLastCheck < 60.) {
-        [self readPurchaserInfoFromCache];
-        return;
-    }
+    return !(self.cachesLastUpdated != nil && timeSinceLastCheck < 60. * 5);
+}
+
+- (void)updateCaches {
+    [self updateCachesWithCompletionBlock:nil];
+}
 
+- (void)updateCachesWithCompletionBlock:(RCReceivePurchaserInfoBlock _Nullable)completion
+{
     self.cachesLastUpdated = [NSDate date];
+    [self updatePurchaserInfoCache:completion];
+    [self updateEntitlementsCache:nil];
```

Comment:
https://github.com/RevenueCat/purchases-ios/pull/70/files#diff-1f8164986a491f4cc448a01a1161e2bbR437 invalidates the cache, but in case there is an error updating entitlements, we don't invalidate. I think we should add `self.cachesLastUpdated = nil;` in https://github.com/RevenueCat/purchases-ios/pull/70/files#diff-1f8164986a491f4cc448a01a1161e2bbR476